algo-builder-tester/test/v6trade-test.ts

```diff
@@ -0,0 +1,401 @@
+    syncAccounts();
+    const dispenseBtcTxParams: typesW.AssetTransferParam = {
+      type: typesW.TransactionType.TransferAsset,
+      sign: typesW.SignType.SecretKey,
+      fromAccount: admin.account,
+      assetID: btcID,
+      payFlags: { totalFee: fee },
+      toAccountAddr: admin.address,
+      amount: dispensedInit, // 1 aBTC
+    };
+    dispenseBtcTxParams.toAccountAddr = alice.address;
+    runtime.executeTx(dispenseBtcTxParams);
+    dispenseBtcTxParams.toAccountAddr = billy.address;
+    runtime.executeTx(dispenseBtcTxParams);
+    const dispenseUsdTxParams: typesW.AssetTransferParam = {
+      type: typesW.TransactionType.TransferAsset,
+      sign: typesW.SignType.SecretKey,
+      fromAccount: admin.account,
+      assetID: usdID,
+      payFlags: { totalFee: fee },
+      toAccountAddr: admin.address,
+      amount: dispensedInit, // 1 aBTC
+    };
+    dispenseUsdTxParams.toAccountAddr = alice.address;
+    runtime.executeTx(dispenseUsdTxParams);
+    dispenseUsdTxParams.toAccountAddr = billy.address;
+    runtime.executeTx(dispenseUsdTxParams);
+    return { aBTC, aUSD };
+  }
+
+  describe("smart contract basic set up", function () {
+    it("check creator", function () {
+      const btcCreator = runtime.getAssetAccount(btcID);
+      const usdCreator = runtime.getAssetAccount(usdID);
+      const SCCreator = runtime.getApp(appID).creator;
+      assert.equal(btcCreator.address, admin.address);
+      assert.equal(usdCreator.address, admin.address);
+      assert.equal(SCCreator, admin.address);
+    });
+    it("assert alice opted into ASA", function () {
+      syncAccounts();
+      const last_msg: Uint8Array = alice.getLocalState(
+        appID,
+        "last_msg"
+      ) as Uint8Array;
+      assert.equal(u8a2Str(last_msg), "OptIn OK.");
+    });
+  });
+  describe("buy/sell with smart contract, static price", function () {
+    let aliceCallParam: typesW.AppCallsParam;
+    let alicePayTxParam: typesW.AssetTransferParam;
+    let aliceCollectTxParam: typesW.AssetTransferParam;
+    this.beforeEach(function () {
+      aliceCallParam = {
+        type: typesW.TransactionType.CallApp,
+        sign: typesW.SignType.SecretKey,
+        fromAccount: alice.account,
+        appID: appID,
+        payFlags: { totalFee: fee },
+        appArgs: [],
+      };
+      alicePayTxParam = {
+        type: typesW.TransactionType.TransferAsset,
+        sign: typesW.SignType.SecretKey,
+        assetID: 0,
+        amount: 0,
+        fromAccount: alice.account,
+        toAccountAddr: admin.address,
+        payFlags: { totalFee: fee },
+      };
+      aliceCollectTxParam = {
+        type: typesW.TransactionType.TransferAsset,
+        sign: typesW.SignType.SecretKey,
+        assetID: 0,
+        amount: 0,
+        fromAccount: admin.account,
+        toAccountAddr: alice.address,
+        payFlags: { totalFee: fee },
+      };
+    });
+
+    function resetInitStatus() {
```

Comment:
There are a few place that mix helper functions in the test, this makes tests hard to read. It would be nice to place all functions before or after tests, and part of test short and neat:
```
describe('...', function() {
    it('...', function() {
        ...
    })
    it('...', function() {
        ...
    })
})
```

Reply:
@bo-abstrlabs @ailisp need some answer here.

Reply:
Pass variables sounds good to me. at begining of describe - I still think if the function is too lengthy it's isn't good. If
```
describe('...', function() {
    beforeAll()
    afterEach()
    it('...', function() {
        ...
    })
    it('...', function() {
        helper1()
        helper2()
        ...
    })
})
```
part is as short as one to two screens, it's very easy to scan read what are being tested

Of course we should also balance the effort to move helpers and productivity to write tests

Reply:
Yes, the productivity. In our code style it would be `{alice,billy,admin} = syncAcc(runtime)` everytime. 
Maybe we can do 

```
function lenthy_helper()     // new
describe('...', function() {
    define_vars(){}
    function wrapped_common_helper(){}     // new
    beforeAll()
    afterEach()
    it('...', function() {
        wrapped_common_helper(){}      // new
        ...
    })
    it('...', function() {
        wrapped_common_helper(){}.     // new
        ...
    })
})

Reply:
@bo-abstrlabs 
So the `syncAcc()` would become two functions `syncAccWithParam(runtime) : { alice, ...} { /*algorithm*/ }` then `syncAcc() { alice... = syncAccWithParam(runtime) }

Reply:
in this case why wrapped_common_helper cannot be defined outside of `describe`? to make `describe` block close to `it` blocks.

Reply:
(repeat & translate from slack)
I don't know why `alice` and `runtime` are defined in `describe` in those examples of `algob`. I guess it's because `mocha` adds a this to each `describe`. If the `wrapped_helper` is defined outside, they cannot are not assign to the value defined within `describe()`.